File: sql_parser.py
import sqlparse
from sqlparse.sql import Identifier, IdentifierList, Parenthesis
from sqlparse.tokens import Keyword, DML, Whitespace

class SQLParser:
    def __init__(self):
        pass
        
    def normalize_sql_definition(self, definition):
        """标准化SQL定义，统一大小写和格式"""
        # 将定义转换为大写
        definition = definition.upper()
        # 标准化空格
        definition = ' '.join(definition.split())
        return definition
        
    def parse_file(self, file_path):
        """解析SQL文件，返回表结构字典"""
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            
        # 解析SQL语句
        statements = sqlparse.parse(content)
        tables = {}
        
        for statement in statements:
            # 只处理CREATE TABLE语句
            if not statement.get_type().upper() == 'CREATE':
                continue
                
            # 获取表名和列定义
            table_name = None
            columns = {}
            indexes = {}
            
            # 遍历语句的token
            for item in statement.tokens:
                if isinstance(item, Identifier):
                    # 获取表名
                    if not table_name:
                        table_name = item.get_name().strip('`')
                        continue
                        
                # 处理括号内的列定义和索引
                if isinstance(item, Parenthesis):
                    # 获取括号内的内容
                    content = item.value.strip('()')
                    # 分割每个定义
                    defs = [d.strip() for d in content.split(',')]
                    
                    for definition in defs:
                        if not definition:
                            continue
                            
                        # 处理索引定义
                        if definition.upper().startswith(('PRIMARY KEY', 'KEY', 'UNIQUE KEY', 'FOREIGN KEY')):
                            index_type = None
                            index_name = None
                            index_columns = None
                            
                            # 解析索引类型和名称
                            parts = definition.split(None, 2)
                            if len(parts) >= 2:
                                if parts[0].upper() == 'PRIMARY':
                                    index_type = 'PRIMARY KEY'
                                    index_name = 'PRIMARY'
                                    index_columns = parts[2].strip('()')
                                else:
                                    index_type = parts[0].upper()
                                    if len(parts) > 2:
                                        index_name = parts[1].strip('`')
                                        index_columns = parts[2].strip('()')
                                        
                                if index_type and index_columns:
                                    indexes[index_name] = {
                                        'type': index_type,
                                        'columns': index_columns
                                    }
                            continue
                            
                        # 处理列定义
                        parts = definition.split(None, 1)
                        if len(parts) >= 2:
                            col_name = parts[0].strip('`')
                            col_type = parts[1].strip()
                            # 存储原始定义和标准化后的定义
                            columns[col_name] = {
                                'raw': col_type,
                                'normalized': self.normalize_sql_definition(col_type)
                            }
                            
            if table_name:
                tables[table_name] = {
                    'columns': columns,
                    'indexes': indexes,
                    'raw_sql': str(statement)
                }
                
        return tables
        
    def compare_tables(self, left_tables, right_tables):
        """比较两个表结构的差异"""
        differences = {
            'added_tables': [],
            'removed_tables': [],
            'modified_tables': {}
        }
        
        # 检查新增和删除的表
        left_table_names = set(left_tables.keys())
        right_table_names = set(right_tables.keys())
        
        differences['added_tables'] = list(right_table_names - left_table_names)
        differences['removed_tables'] = list(left_table_names - right_table_names)
        
        # 检查修改的表
        common_tables = left_table_names & right_table_names
        for table_name in common_tables:
            left_cols = left_tables[table_name]['columns']
            right_cols = right_tables[table_name]['columns']
            left_indexes = left_tables[table_name]['indexes']
            right_indexes = right_tables[table_name]['indexes']
            
            table_diffs = {}
            
            # 比较列定义
            if left_cols != right_cols:
                table_diffs['columns'] = {
                    'added_columns': {col: defn['raw'] for col, defn in right_cols.items() if col not in left_cols},
                    'removed_columns': {col: defn['raw'] for col, defn in left_cols.items() if col not in right_cols},
                    'modified_columns': {
                        col: {'left': left_cols[col]['raw'], 'right': right_cols[col]['raw']}
                        for col in set(left_cols.keys()) & set(right_cols.keys())
                        if left_cols[col]['normalized'] != right_cols[col]['normalized']
                    }
                }
            
            # 比较索引定义
            if left_indexes != right_indexes:
                table_diffs['indexes'] = {
                    'added_indexes': {idx: defn for idx, defn in right_indexes.items() if idx not in left_indexes},
                    'removed_indexes': {idx: defn for idx, defn in left_indexes.items() if idx not in right_indexes},
                    'modified_indexes': {
                        idx: {'left': left_indexes[idx], 'right': right_indexes[idx]}
                        for idx in set(left_indexes.keys()) & set(right_indexes.keys())
                        if left_indexes[idx] != right_indexes[idx]
                    }
                }
            
            if table_diffs:
                differences['modified_tables'][table_name] = table_diffs
                
        return differences 

File: src/main/sql-parser.ts
import * as fs from 'fs';
import { Tables, TableDifferences } from './types';

export class SQLParser {
  parseFile(filePath: string): Tables {
    const content = fs.readFileSync(filePath, 'utf-8');
    return this.parseSQL(content);
  }

  parseSQL(sql: string): Tables {
    const tables: Tables = {};
    
    // 简单的SQL解析逻辑
    const createTableRegex = /CREATE\s+TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?`?(\w+)`?\s*\(([\s\S]*?)\)\s*;/gi;
    let match;
    
    while ((match = createTableRegex.exec(sql)) !== null) {
      const tableName = match[1];
      const tableBody = match[2];
      
      tables[tableName] = this.parseTableBody(tableBody);
    }
    
    return tables;
  }

  private parseTableBody(body: string): any {
    const columns: { [key: string]: any } = {};
    const indexes: { [key: string]: any } = {};
    
    const lines = body.split('\n').map(line => line.trim()).filter(line => line);
    
    for (const line of lines) {
      if (line.startsWith('`') || /^\w+\s+/.test(line)) {
        // 解析列定义
        const columnMatch = line.match(/`?(\w+)`?\s+([^,\s]+)(?:\(([^)]+)\))?\s*(NOT\s+NULL|NULL)?\s*(DEFAULT\s+[^,\s]+)?\s*(COMMENT\s+'[^']*')?/i);
        if (columnMatch) {
          const columnName = columnMatch[1];
          const columnType = columnMatch[2];
          const length = columnMatch[3];
          const nullable = !columnMatch[4] || columnMatch[4].toUpperCase() !== 'NOT NULL';
          const defaultValue = columnMatch[5];
          const comment = columnMatch[6];
          
          columns[columnName] = {
            raw: line,
            type: columnType,
            length: length,
            nullable: nullable,
            default: defaultValue,
            comment: comment
          };
        }
      } else if (line.toUpperCase().includes('KEY') || line.toUpperCase().includes('INDEX')) {
        // 解析索引定义
        const indexMatch = line.match(/(?:UNIQUE\s+)?(?:KEY|INDEX)\s+`?(\w+)`?\s*\(([^)]+)\)/i);
        if (indexMatch) {
          const indexName = indexMatch[1];
          const indexColumns = indexMatch[2];
          const isUnique = line.toUpperCase().includes('UNIQUE');
          
          indexes[indexName] = {
            type: isUnique ? 'UNIQUE' : 'INDEX',
            columns: indexColumns,
            unique: isUnique
          };
        }
      }
    }
    
    return { columns, indexes };
  }

  compareTables(leftTables: Tables, rightTables: Tables): TableDifferences {
    const differences: TableDifferences = {
      modifiedTables: {},
      addedTables: [],
      removedTables: []
    };
    
    const allTableNames = new Set([
      ...Object.keys(leftTables),
      ...Object.keys(rightTables)
    ]);
    
    for (const tableName of allTableNames) {
      const leftTable = leftTables[tableName];
      const rightTable = rightTables[tableName];
      
      if (!leftTable) {
        differences.addedTables.push(tableName);
      } else if (!rightTable) {
        differences.removedTables.push(tableName);
      } else {
        const tableDiff = this.compareTableStructure(leftTable, rightTable);
        if (tableDiff) {
          differences.modifiedTables[tableName] = tableDiff;
        }
      }
    }
    
    return differences;
  }

  private compareTableStructure(leftTable: any, rightTable: any): any {
    const differences: any = {};
    
    // 比较列
    const columnDiff = this.compareColumns(leftTable.columns, rightTable.columns);
    if (columnDiff) {
      differences.columns = columnDiff;
    }
    
    // 比较索引
    const indexDiff = this.compareIndexes(leftTable.indexes, rightTable.indexes);
    if (indexDiff) {
      differences.indexes = indexDiff;
    }
    
    return Object.keys(differences).length > 0 ? differences : null;
  }

  private compareColumns(leftColumns: any, rightColumns: any): any {
    const differences: any = {
      addedColumns: {},
      removedColumns: {},
      modifiedColumns: {}
    };
    
    const allColumnNames = new Set([
      ...Object.keys(leftColumns),
      ...Object.keys(rightColumns)
    ]);
    
    for (const columnName of allColumnNames) {
      const leftColumn = leftColumns[columnName];
      const rightColumn = rightColumns[columnName];
      
      if (!leftColumn) {
        differences.addedColumns[columnName] = rightColumn;
      } else if (!rightColumn) {
        differences.removedColumns[columnName] = leftColumn;
      } else if (JSON.stringify(leftColumn) !== JSON.stringify(rightColumn)) {
        differences.modifiedColumns[columnName] = {
          left: leftColumn,
          right: rightColumn
        };
      }
    }
    
    return Object.keys(differences.addedColumns).length > 0 ||
           Object.keys(differences.removedColumns).length > 0 ||
           Object.keys(differences.modifiedColumns).length > 0 ? differences : null;
  }

  private compareIndexes(leftIndexes: any, rightIndexes: any): any {
    const differences: any = {
      addedIndexes: [],
      removedIndexes: [],
      modifiedIndexes: []
    };
    
    const allIndexNames = new Set([
      ...Object.keys(leftIndexes),
      ...Object.keys(rightIndexes)
    ]);
    
    for (const indexName of allIndexNames) {
      const leftIndex = leftIndexes[indexName];
      const rightIndex = rightIndexes[indexName];
      
      if (!leftIndex) {
        differences.addedIndexes.push(indexName);
      } else if (!rightIndex) {
        differences.removedIndexes.push(indexName);
      } else if (JSON.stringify(leftIndex) !== JSON.stringify(rightIndex)) {
        differences.modifiedIndexes.push(indexName);
      }
    }
    
    return differences.addedIndexes.length > 0 ||
           differences.removedIndexes.length > 0 ||
           differences.modifiedIndexes.length > 0 ? differences : null;
  }
} 